tempesta_fw/sock_srv.c

```diff
@@ -1248,262 +1476,513 @@ static int
-		tfw_sock_srv_del_conns(srv);
-		tfw_server_destroy(srv);
+	tfw_cfgop_cleanup_srv_cfgs();
+
+	if (tfw_runstate_is_reconfig()) {
+		tfw_sg_for_each_srv(true, tfw_sock_srv_del_conns);
+
+		return;
 	}
-	list_for_each_entry_safe(srv, tmp, &tfw_cfg_out_slst, list) {
-		list_del(&srv->list);
+
+	/* Release active configuration. */
+	tfw_sock_srv_delete_all_conns();
+	tfw_sg_release_all();
+
+	TFW_DBG2("clean orphaned servers\n");
+	/* Release structures from previous reconfigs. */
+	list_for_each_entry_safe(srv, srv_tmp, &orphan_srvs, list) {
 		tfw_sock_srv_del_conns(srv);
 		tfw_server_destroy(srv);
 	}
+	TFW_DBG2("clean orphaned groups\n");
+	list_for_each_entry_safe(sg, sg_tmp, &orphan_sgs, list) {
+		__tfw_sg_for_each_srv(sg, tfw_sock_srv_del_conns);
+		tfw_sg_release(sg);
+	}
+	INIT_LIST_HEAD(&orphan_sgs);
+}
 
-	tfw_cfg_slstsz = tfw_cfg_out_slstsz = 0;
-	tfw_cfg_is_set = cfg_is_set_empty;
 
-	tfw_sock_srv_delete_all_conns();
-	tfw_sg_release_all();
+static int
+tfw_sock_srv_cfgstart(void)
+{
+	int r;
+
+	INIT_LIST_HEAD(&sg_cfg_list);
+	if ((r = tfw_cfgop_new_sg_cfg_default()))
+		return r;
+
+	tfw_cfg_sg = tfw_cfg_sg_def;
+
+	return 0;
 }
 
 static int
-tfw_sock_srv_start(void)
+tfw_sock_srv_cfgend(void)
 {
-	int ret;
-
+	TfwSrvGroup *sg;
+	int r;
 	/*
 	 * The group "default" is created implicitly, and only when
 	 * a server outside of any group is found in the configuration.
 	 */
-	if (tfw_cfg_out_slstsz) {
-		tfw_cfg_out_sg = tfw_sg_new("default", GFP_KERNEL);
-		if (!tfw_cfg_out_sg) {
-			TFW_ERR_NL("Unable to add default server group\n");
-			return -EINVAL;
+	if (!tfw_cfg_sg_def->parsed_sg->srv_n) {
+		tfw_cfgop_cleanup_srv_cfg(tfw_cfg_sg_def);
+		tfw_cfg_sg_def = NULL;
+		return 0;
+	}
+
+	if ((r = tfw_cfgop_setup_srv_group(tfw_cfg_sg_def)))
+		return r;
+
+	sg = tfw_cfg_sg_def->orig_sg ? : tfw_cfg_sg_def->parsed_sg;
+	if (tfw_sg_add(sg)) {
+		TFW_ERR_NL("Unable to register implicit 'default' group\n");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int
+tfw_cfgop_start_srv(TfwServer *srv)
+{
+	int r;
+
+	TFW_DBG_ADDR("start server", &srv->addr);
+
+	if ((r = tfw_sock_srv_add_conns(srv)))
+		return r;
+	if ((r = tfw_apm_add_srv(srv)))
+		return r;
+	if ((r = tfw_sock_srv_connect_srv(srv)))
+		return r;
+
+	return 0;
+}
+
+static int
+tfw_cfgop_update_srv(TfwServer *orig_srv, TfwCfgSrvGroup *sg_cfg)
+{
+	TfwServer *srv = NULL;
+	int r;
+
+	list_for_each_entry(srv, &sg_cfg->parsed_sg->srv_list, list) {
+		if (tfw_addr_eq(&srv->addr, &orig_srv->addr))
+			break;
+	}
+
+	if (!srv)
+		return -EINVAL;
+
+	TFW_DBG_ADDR("Update server options", &srv->addr);
+
+	orig_srv->weight = srv->weight;
+
+	if (orig_srv->conn_n < srv->conn_n) {
+		if ((r = tfw_sock_srv_append_conns_n(
+			     srv, srv->conn_n - orig_srv->conn_n)))
+			return r;
+
+		orig_srv->conn_n = srv->conn_n;
+	}
+	else if (orig_srv->conn_n > srv->conn_n)
+	{
+		/*
+		 * TODO: shrink connections: disconnect connection and remove
+		 * it from server's conn_list
+		 */
```

Comment:
It's a bad thing to leave server connection open - it could lead to serious memory leakage. Why don't you use `tfw_sock_srv_disconnect()`?

Reply:
The function was called later in one loop for all removed servers. I've reworked this to call disconnect() as soon as possible.